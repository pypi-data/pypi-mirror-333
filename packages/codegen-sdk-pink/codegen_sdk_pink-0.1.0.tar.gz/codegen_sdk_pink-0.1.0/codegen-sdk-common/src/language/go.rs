use super::Language;
lazy_static! {
    pub static ref Go: Language = Language::new(
        "go",
        "Go",
        tree_sitter_go::NODE_TYPES,
        &["go"],
        tree_sitter_go::LANGUAGE.into(),
        tree_sitter_go::TAGS_QUERY,
    )
    .unwrap();
}
