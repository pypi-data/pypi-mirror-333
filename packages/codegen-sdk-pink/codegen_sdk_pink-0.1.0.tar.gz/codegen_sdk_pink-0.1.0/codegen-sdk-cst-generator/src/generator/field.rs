// Representation of a field of a node. Can be optional, required, or, multiple
use std::collections::HashSet;

#[double]
use codegen_sdk_common::language::Language;
use codegen_sdk_common::{
    naming::{field_name_getter, normalize_field_name, normalize_type_name},
    parser::{FieldDefinition, TypeDefinition},
};
use mockall_double::double;
use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::parse_quote;

use crate::Config;

#[derive(Debug)]
pub struct Field<'a> {
    raw: FieldDefinition,
    name: String,
    node_name: String,
    language: &'a Language,
    config: Config,
}

impl<'a> Field<'a> {
    pub fn new(
        node_name: &str,
        name: &str,
        raw: FieldDefinition,
        language: &'a Language,
        config: Config,
    ) -> Self {
        Self {
            node_name: node_name.to_string(),
            name: name.to_string(),
            raw,
            language,
            config,
        }
    }
    fn field_id(&self) -> u16 {
        self.language.field_id(&self.name).unwrap().into()
    }
    pub fn name(&self) -> String {
        normalize_field_name(&self.name)
    }
    pub fn normalized_name(&self) -> String {
        normalize_type_name(&self.name, true)
    }
    pub fn types(&self) -> Vec<&TypeDefinition> {
        self.raw.types.iter().collect()
    }
    pub fn type_name(&self) -> String {
        let types = self.types();
        if types.len() == 1 {
            normalize_type_name(&types[0].type_name, types[0].named)
        } else {
            format!("{}{}", self.node_name, self.normalized_name())
        }
    }
    pub fn type_name_ref(&self, subenums: &HashSet<&String>) -> TokenStream {
        let types = self.types();
        // We return a direct reference if the there is only one type and it is not a subenum
        if types.len() == 1 {
            let ident = format_ident!("{}", types[0].normalize());
            if subenums.contains(&types[0].type_name) {
                let ident = format_ident!("{}Ref", ident);
                quote! {#ident}
            } else {
                quote! {
                    &#ident
                }
            }
        } else {
            let ident = format_ident!("{}{}Ref", self.node_name, self.normalized_name());
            quote! {
                #ident
            }
        }
    }

    pub fn get_constructor_field(&self) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        let original_name = &self.name;
        let converted_type_name = format_ident!("{}", self.type_name());
        if self.raw.multiple {
            quote! {
                let #field_name_ident= get_multiple_children_by_field_name::<NodeTypes<'db>, #converted_type_name<'db>>(context, &node, #original_name)?;
                for child in #field_name_ident.iter().cloned() {
                    ids.push(child);
                }
            }
        } else if !self.raw.required {
            quote! {
                let #field_name_ident = get_optional_child_by_field_name::<NodeTypes<'db>, #converted_type_name<'db>>(context, &node, #original_name)?;
                if let Some(child) = #field_name_ident.clone() {
                    ids.push(child);
                }
            }
        } else {
            quote! {
                let #field_name_ident = get_child_by_field_name::<NodeTypes<'db>, #converted_type_name<'db>>(context, &node, #original_name)?;
                ids.push(#field_name_ident.clone());
            }
        }
    }
    pub fn get_convert_child(&self, convert_children: bool) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        if convert_children {
            if self.raw.multiple {
                quote! {
                    context.get(child).unwrap().as_ref().try_into().unwrap()
                }
            } else if !self.raw.required {
                quote! {
                    context.get(child).unwrap().as_ref().try_into().unwrap()
                }
            } else {
                quote! {
                    context.get(&self.#field_name_ident).unwrap().as_ref().try_into().unwrap()
                }
            }
        } else if self.raw.multiple || !self.raw.required {
            quote! {
                child
            }
        } else {
            quote! {
                self.#field_name_ident.as_ref()
            }
        }
    }
    pub fn get_children_field(&self, convert_children: bool) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        let convert_child = self.get_convert_child(convert_children);

        if self.raw.multiple {
            quote! {
                children.extend(self.#field_name_ident.iter().map(|child| #convert_child));
            }
        } else if self.raw.required {
            quote! {
                children.push(#convert_child);
            }
        } else {
            quote! {
                if let Some(child) = self.#field_name_ident.as_ref() {
                    children.push(#convert_child);
                }
            }
        }
    }
    pub fn get_children_by_field_name_field(&self, convert_children: bool) -> TokenStream {
        let field_name = &self.name;
        let field_name_ident = format_ident!("{}", self.name());
        let convert_child = self.get_convert_child(convert_children);

        if self.raw.multiple {
            quote! {
                #field_name => self.#field_name_ident.iter().map(|child| #convert_child).collect()
            }
        } else if self.raw.required {
            quote! {
                #field_name => vec![#convert_child]
            }
        } else {
            quote! {
                #field_name => self.#field_name_ident.as_ref().iter().map(|child| #convert_child).collect()
            }
        }
    }
    pub fn get_children_by_field_id_field(&self, convert_children: bool) -> TokenStream {
        let field_id = self.field_id();
        let field_name_ident = format_ident!("{}", self.name());
        let convert_child = self.get_convert_child(convert_children);

        if self.raw.multiple {
            quote! {
                #field_id => self.#field_name_ident.iter().map(|child| #convert_child).collect()
            }
        } else if self.raw.required {
            quote! {
                #field_id => vec![#convert_child]
            }
        } else {
            quote! {
                #field_id => self.#field_name_ident.as_ref().iter().map(|child| #convert_child).collect()
            }
        }
    }
    pub fn get_struct_field(&self) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        let bounds = if self.config.serialize {
            quote! {
                #[rkyv(omit_bounds)]
            }
        } else {
            quote! {}
        };
        if self.raw.multiple {
            quote! {
                #bounds
                pub #field_name_ident: Vec<indextree::NodeId>
            }
        } else if !self.raw.required {
            quote! {
                #bounds
                pub #field_name_ident: Option<indextree::NodeId>
            }
        } else {
            quote! {
                #bounds
                pub #field_name_ident: indextree::NodeId
            }
        }
    }
    pub fn get_field_getter(&self, subenums: &HashSet<&String>) -> syn::ImplItemFn {
        let function_name = format_ident!("{}", field_name_getter(&self.name()));
        let field_name_ident = format_ident!("{}", self.name());
        let converted_type_name = self.type_name_ref(subenums);
        if self.raw.multiple {
            parse_quote! {
                pub fn #function_name(&self, tree: &'db Tree<NodeTypes<'db>>) -> Vec<#converted_type_name<'db>> {
                    self.#field_name_ident.iter().map(|id| tree.get(id).unwrap().as_ref().try_into().unwrap()).collect()
                }
            }
        } else if !self.raw.required {
            parse_quote! {
                pub fn #function_name(&self, tree: &'db Tree<NodeTypes<'db>>) -> Option<#converted_type_name<'db>> {
                    if let Some(id) = self.#field_name_ident {
                        Some(tree.get(&id).unwrap().as_ref().try_into().unwrap())
                    } else {
                        None
                    }
                }
            }
        } else {
            parse_quote! {
                pub fn #function_name(&self, tree: &'db Tree<NodeTypes<'db>>) -> #converted_type_name<'db> {
                    tree.get(&self.#field_name_ident).unwrap().as_ref().try_into().unwrap()
                }
            }
        }
    }

    pub fn is_optional(&self) -> bool {
        !self.raw.required
    }
    pub fn is_multiple(&self) -> bool {
        self.raw.multiple
    }
}

#[cfg(test)]
mod tests {
    use codegen_sdk_common::parser::TypeDefinition;

    use super::*;
    use crate::test_util::{get_language_no_nodes, snapshot_tokens};
    fn create_test_field_definition(name: &str, multiple: bool, required: bool) -> FieldDefinition {
        FieldDefinition {
            types: vec![TypeDefinition {
                type_name: name.to_string(),
                named: true,
            }],
            multiple,
            required,
        }
    }
    fn create_test_field_definition_variants(
        name: &Vec<String>,
        multiple: bool,
        required: bool,
    ) -> FieldDefinition {
        FieldDefinition {
            types: name
                .iter()
                .map(|n| TypeDefinition {
                    type_name: n.to_string(),
                    named: true,
                })
                .collect(),
            multiple,
            required,
        }
    }

    #[test]
    fn test_field_normalized_name() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "node",
            "field",
            field_definition,
            &language,
            Config::default(),
        );
        assert_eq!(field.normalized_name(), "Field");
    }

    #[test]
    fn test_field_types() {
        let field_definition = create_test_field_definition_variants(
            &vec!["type_a".to_string(), "type_b".to_string()],
            false,
            true,
        );
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition.clone(),
            &language,
            Config::default(),
        );
        assert_eq!(
            field.types(),
            field_definition.types.iter().collect::<Vec<_>>()
        );
    }

    #[test]
    fn test_field_type_name() {
        let field_definition = create_test_field_definition_variants(
            &vec!["test_type".to_string(), "test_type".to_string()],
            false,
            true,
        );
        let language = get_language_no_nodes();
        let field = Field::new(
            "Node",
            "field",
            field_definition,
            &language,
            Config::default(),
        );
        assert_eq!(field.type_name(), "NodeField");
    }

    #[test]
    fn test_get_struct_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&field.get_struct_field()));

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&optional_field.get_struct_field()));

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&multiple_field.get_struct_field()));
    }

    #[test]
    fn test_get_constructor_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&field.get_constructor_field()));

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&optional_field.get_constructor_field()));

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&multiple_field.get_constructor_field()));
    }

    #[test]
    fn test_get_children_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            field.get_children_field(true).to_string(),
            quote!(children.push(context.get(&self.test_field).unwrap().as_ref().try_into().unwrap());).to_string()
        );

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            optional_field.get_children_field(true).to_string(),
            quote!(if let Some(child) = self.test_field.as_ref() {
                children.push(context.get(child).unwrap().as_ref().try_into().unwrap());
            })
            .to_string()
        );

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            multiple_field.get_children_field(true).to_string(),
            quote!(children.extend(self.test_field.iter().map(|child| context.get(child).unwrap().as_ref().try_into().unwrap()));).to_string()
        );
    }

    #[test]
    fn test_get_children_by_field_name_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            field.get_children_by_field_name_field(true).to_string(),
            quote!("test_field" => vec![context.get(&self.test_field).unwrap().as_ref().try_into().unwrap()]).to_string()
        );

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            optional_field.get_children_by_field_name_field(true).to_string(),
            quote!("test_field" => self.test_field.as_ref().iter().map(|child| context.get(child).unwrap().as_ref().try_into().unwrap()).collect()).to_string()
        );

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            multiple_field.get_children_by_field_name_field(true).to_string(),
            quote!("test_field" => self.test_field.iter().map(|child| context.get(child).unwrap().as_ref().try_into().unwrap()).collect()).to_string()
        );
    }
}
