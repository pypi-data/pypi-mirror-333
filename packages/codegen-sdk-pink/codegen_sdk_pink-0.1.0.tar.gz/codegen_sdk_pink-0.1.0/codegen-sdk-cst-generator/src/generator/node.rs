// Representation of a tree-sitter node without variants
use std::{
    collections::{HashMap, HashSet},
    sync::Arc,
};

#[double]
use codegen_sdk_common::language::Language;
use codegen_sdk_common::{naming::normalize_type_name, parser::TypeDefinition};
use mockall_double::double;
use proc_macro2::TokenStream;
use quote::{format_ident, quote};

use super::field::Field;
use crate::{
    Config,
    generator::utils::{get_comment_type, get_serialize_bounds},
};

#[derive(Debug)]
pub struct Node<'a> {
    raw: Arc<codegen_sdk_common::parser::Node>,
    pub subenums: Vec<String>,
    pub fields: Vec<Field<'a>>,
    language: &'a Language,
    config: Config,
    normalized_name: String,
}
impl<'a> Node<'a> {
    pub fn new(
        raw: Arc<codegen_sdk_common::parser::Node>,
        language: &'a Language,
        config: Config,
    ) -> Self {
        let mut fields = Vec::new();
        let normalized_name = normalize_type_name(&raw.type_name, raw.named);
        if let Some(raw_fields) = &raw.fields {
            for (name, field) in raw_fields.fields.iter() {
                fields.push(Field::new(
                    &normalized_name,
                    name,
                    field.clone(),
                    language,
                    config.clone(),
                ));
            }
        }
        fields.sort_by_key(|f| f.normalized_name().clone());
        Node {
            raw,
            subenums: Vec::new(),
            fields,
            language,
            config: config,
            normalized_name,
        }
    }
    pub fn kind(&self) -> &str {
        &self.raw.type_name
    }
    pub fn kind_id(&self) -> u16 {
        self.language.kind_id(&self.raw.type_name, self.raw.named)
    }
    pub fn normalize_name(&self) -> String {
        self.normalized_name.clone()
    }
    pub fn type_definition(&self) -> TypeDefinition {
        TypeDefinition {
            type_name: self.raw.type_name.clone(),
            named: self.raw.named,
        }
    }
    pub fn add_subenum(&mut self, subenum: String) {
        if !self.subenums.contains(&subenum) {
            self.subenums.push(subenum);
        }
    }
    pub fn get_enum_tokens(
        &self,
        subenum_name_map: &HashMap<String, String>,
        is_ref: bool,
    ) -> TokenStream {
        let name = format_ident!("{}", self.normalize_name());
        let value = if is_ref {
            quote! { &'db1 #name }
        } else {
            quote! { #name }
        };
        let subenum_names = &self
            .subenums
            .iter()
            .map(|s| subenum_name_map.get(s).unwrap_or(&s))
            .map(|s| {
                if is_ref {
                    format_ident!("{}Ref", s)
                } else {
                    format_ident!("{}", s)
                }
            })
            .collect::<Vec<_>>();
        if subenum_names.is_empty() {
            quote! {
                #name(#value<'db1>)
            }
        } else {
            quote! {
                #[subenum(#(#subenum_names), *)]
                #name(#value<'db1>)
            }
        }
    }
    pub fn get_children_names(&self) -> Vec<TypeDefinition> {
        let mut children_names = vec![];
        let comment = get_comment_type();
        if let Some(children) = &self.raw.children {
            children_names.extend(children.types.iter().cloned());
        }
        for field in &self.fields {
            children_names.extend(field.types().into_iter().cloned());
        }
        if children_names.len() > 0 && !children_names.contains(&comment) {
            children_names.push(comment);
        }
        children_names.sort();
        children_names.dedup();
        children_names
    }
    pub fn children_struct_name(&self) -> String {
        let children_names = self.get_children_names();
        match children_names.len() {
            0 => "Self".to_string(),
            1 => normalize_type_name(&children_names[0].type_name, children_names[0].named),
            _ => format!("{}Children", self.normalize_name()),
        }
    }
    pub fn has_children(&self) -> bool {
        self.raw.children.is_some()
    }
    fn get_children_field(&self) -> TokenStream {
        if self.has_children() {
            let bounds = if self.config.serialize {
                quote! {
                    #[rkyv(omit_bounds)]
                }
            } else {
                quote! {}
            };
            quote! {
                #bounds
                pub _children: Vec<indextree::NodeId>,
            }
        } else {
            quote! {}
        }
    }

    pub fn get_struct_tokens(&self, subenums: &HashSet<&String>) -> TokenStream {
        let constructor = self.get_constructor();
        let struct_fields = self
            .fields
            .iter()
            .map(|f| f.get_struct_field())
            .collect::<Vec<_>>();
        let children_field = self.get_children_field();
        let name = format_ident!("{}", self.normalize_name());
        let trait_impls = self.get_trait_implementations(subenums);
        let derives = if self.config.serialize {
            let serialize_bounds = get_serialize_bounds();
            quote! {
                #[derive(Debug, Deserialize, Clone, Archive, Serialize, Eq, PartialEq, salsa::Update)]
                #serialize_bounds
            }
        } else {
            quote! {
                #[derive(Debug, Eq, PartialEq, Clone, salsa::Update)]

            }
        };
        quote! {
            #derives
            pub struct #name<'db> {
                id: CSTNodeId<'db>,
                file_id: FileNodeId,
                start_byte: usize,
                end_byte: usize,
                start_position: Point<'db>,
                end_position: Point<'db>,
                #[debug(ignore)]
                buffer: Arc<Bytes>,
                is_error: bool,
                #children_field
                #(#struct_fields),*
            }
            #constructor
            #trait_impls
        }
    }
    fn get_children_constructor(&self) -> TokenStream {
        if self.has_children() {
            let children_type_name = format_ident!("{}", self.children_struct_name());
            quote! {
                let _children = named_children_without_field_names::<NodeTypes<'db>, #children_type_name<'db>>(context, node)?;
                for child in _children.iter().cloned() {
                    ids.push(child);
                }
            }
        } else {
            quote! {}
        }
    }
    pub fn get_constructor(&self) -> TokenStream {
        let name = format_ident!("{}", self.normalize_name());
        let mut constructor_fields = Vec::new();
        let mut constructor_names = Vec::new();
        for field in &self.fields {
            constructor_fields.push(field.get_constructor_field());
            constructor_names.push(format_ident!("{}", field.name()));
        }
        constructor_fields.push(self.get_children_constructor());
        if self.has_children() {
            constructor_names.push(format_ident!("_children"));
        }

        quote! {
            impl<'db> FromNode<'db, NodeTypes<'db>> for #name<'db> {
                fn from_node(context: &mut ParseContext<'db, NodeTypes<'db>>, node: tree_sitter::Node) -> Result<(Self, Vec<indextree::NodeId>), ParseError> {
                    let start_position = Point::from(context.db, node.start_position());
                    let end_position = Point::from(context.db, node.end_position());
                    let id = CSTNodeId::new(context.db, context.file_id, node.id(), context.root);
                    let mut ids = Vec::new();
                    #(#constructor_fields)*
                    Ok((Self {
                        start_byte: node.start_byte(),
                        end_byte: node.end_byte(),
                        start_position: start_position,
                        end_position: end_position,
                        buffer: context.buffer.clone(),
                        is_error: node.is_error(),
                        id,
                        file_id: context.file_id.clone(),
                        #(#constructor_names),*
                }, ids))
                }
            }
        }
    }
    fn get_children_impl(&self) -> TokenStream {
        let name = format_ident!("{}", self.normalize_name());

        let children_type_name = self.children_struct_name();
        let children_type_generic;
        if children_type_name != "Self" {
            let children_type_ident = format_ident!("{}Ref", children_type_name);
            children_type_generic = quote! {#children_type_ident<'db2>};
        } else {
            children_type_generic = quote! {#name<'db2>};
        }

        let children_field = self.get_children_field_impl();
        let children_by_field_name = self.get_children_by_field_name_impl();
        let children_by_field_id = self.get_children_by_field_id_impl();
        quote! {
            impl<'db1> HasChildren<'db1, NodeTypes<'db1>> for #name<'db1> {
                type Child<'db2> = #children_type_generic where Self: 'db2;
                #children_field
                #children_by_field_name
                #children_by_field_id
            }
        }
    }
    pub fn get_trait_implementations(&self, subenums: &HashSet<&String>) -> TokenStream {
        let name = format_ident!("{}", self.normalize_name());
        let children_impl = self.get_children_impl();
        let getters = self
            .fields
            .iter()
            .map(|f| f.get_field_getter(subenums))
            .collect::<Vec<_>>();
        let kind_name = &self.raw.type_name;
        let is_named = self.raw.named;
        let kind_id = self.kind_id();
        quote! {
            impl<'db> #name<'db> {
                const KIND_NAME: &'static str = #kind_name;
                #(#getters)*
            }
            impl<'db> CSTNode<'db> for #name<'db> {
                fn kind_name(&self) -> &str {
                    &Self::KIND_NAME
                }
                fn start_byte(&self) -> usize {
                    self.start_byte
                }
                fn end_byte(&self) -> usize {
                    self.end_byte
                }
                fn start_position(&self) -> Point<'db> {
                    self.start_position
                }
                fn end_position(&self) -> Point<'db> {
                    self.end_position
                }
                fn buffer(&self) -> &Bytes {
                    &self.buffer
                }
                fn kind_id(&self) -> u16 {
                    #kind_id
                }
                fn is_error(&self) -> bool {
                    self.is_error
                }
                fn is_named(&self) -> bool {
                    #is_named
                }
                fn id(&self) -> CSTNodeId<'db> {
                    self.id
                }
                fn file_id(&self) -> FileNodeId {
                    self.file_id
                }
            }
            #children_impl
            impl<'db> std::hash::Hash for #name<'db> {
                fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
                    self.id.hash(state);
                }
            }
            impl<'db> PartialOrd for #name<'db> {
                fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
                    Some(self.cmp(other))
                }
            }
            impl<'db> Ord for #name<'db> {
                fn cmp(&self, other: &Self) -> std::cmp::Ordering {
                    let res = self.start_byte().cmp(&other.start_byte());
                    if res == std::cmp::Ordering::Equal {
                        self.end_byte().cmp(&other.end_byte())
                    } else {
                        res
                    }
                }
            }

        }
    }
    fn get_children_field_impl(&self) -> TokenStream {
        let mut children_fields = Vec::new();
        let num_children = self.get_children_names().len();
        if num_children == 0 {
            return quote! {
                fn children<'db2>(&'db2 self, context: &'db2 Tree< NodeTypes<'db2>>) -> Vec<Self::Child<'db2>> {
                    vec![]
                }
            };
        }
        let convert_children = num_children > 1;
        for field in &self.fields {
            children_fields.push(field.get_children_field(convert_children));
        }

        let children_init = if self.has_children() {
            quote! {
                self._children.iter().map(|c| context.get(c).unwrap().as_ref().try_into().unwrap()).collect()
            }
        } else {
            quote! {
                vec![]
            }
        };
        quote! {
            fn children<'db2>(&'db2 self, context: &'db2 Tree< NodeTypes<'db2>>) -> Vec<Self::Child<'db2>> {
                let mut children: Vec<Self::Child<'db2>> = #children_init;
                #(#children_fields;)*
                children.sort_by_key(|c| c.start_byte());
                children
            }
        }
    }
    fn get_children_by_field_name_impl(&self) -> TokenStream {
        let convert_children = self.get_children_names().len() > 1;
        let field_matches = self
            .fields
            .iter()
            .map(|f| f.get_children_by_field_name_field(convert_children))
            .collect::<Vec<_>>();

        quote! {
            fn children_by_field_name<'db2>(&'db2 self, context: &'db2 Tree< NodeTypes<'db2>>, field_name: &str) -> Vec<Self::Child<'db2>> {
                match field_name {
                    #(#field_matches,)*
                    _ => vec![],
                }
            }
        }
    }
    fn get_children_by_field_id_impl(&self) -> TokenStream {
        let convert_children = self.get_children_names().len() > 1;
        let field_matches = self
            .fields
            .iter()
            .map(|f| f.get_children_by_field_id_field(convert_children))
            .collect::<Vec<_>>();

        quote! {
            fn children_by_field_id<'db2>(&'db2 self, context: &'db2 Tree< NodeTypes<'db2>>, field_id: u16) -> Vec<Self::Child<'db2>> {
                match field_id {
                    #(#field_matches,)*
                    _ => vec![],
                }
            }
        }
    }
    pub fn get_field_for_field_name(&self, field_name: &str) -> Option<&Field<'a>> {
        self.fields.iter().find(|f| f.name() == field_name)
    }
}
#[cfg(test)]
mod tests {
    use codegen_sdk_common::parser::{FieldDefinition, Fields, TypeDefinition};

    use super::*;
    use crate::test_util::{get_language, get_language_no_nodes, snapshot_tokens};

    fn create_test_node(name: &str) -> codegen_sdk_common::parser::Node {
        codegen_sdk_common::parser::Node {
            type_name: name.to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        }
    }

    fn create_test_node_with_fields(
        name: &str,
        fields: Vec<(&str, FieldDefinition)>,
    ) -> codegen_sdk_common::parser::Node {
        codegen_sdk_common::parser::Node {
            type_name: name.to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: Some(Fields {
                fields: fields
                    .into_iter()
                    .map(|(k, v)| (k.to_string(), v))
                    .collect(),
            }),
            children: None,
        }
    }

    fn create_test_node_with_children(
        name: &str,
        child_types: Vec<&str>,
    ) -> codegen_sdk_common::parser::Node {
        codegen_sdk_common::parser::Node {
            type_name: name.to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: Some(codegen_sdk_common::parser::Children {
                types: child_types
                    .into_iter()
                    .map(|t| TypeDefinition {
                        type_name: t.to_string(),
                        named: true,
                    })
                    .collect(),
                multiple: false,
                required: true,
            }),
        }
    }

    #[test_log::test]
    fn test_get_enum_tokens() {
        let base_node = create_test_node("test");
        let language = get_language_no_nodes();
        let mut node = Node::new(Arc::new(base_node), &language, Config::default());
        let mut subenum_name_map = HashMap::new();
        for subenum in &node.subenums {
            subenum_name_map.insert(subenum.clone(), normalize_type_name(subenum, true));
        }
        let tokens = node.get_enum_tokens(&subenum_name_map, false);
        insta::assert_debug_snapshot!(snapshot_tokens(&tokens));

        node.add_subenum("subenum".to_string());
        let tokens = node.get_enum_tokens(&subenum_name_map, false);
        insta::assert_debug_snapshot!(snapshot_tokens(&tokens));
    }

    #[test]
    fn test_get_struct_tokens_simple() {
        let raw_node = create_test_node("test_node");
        let language = get_language_no_nodes();
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(&node.get_struct_tokens(&HashSet::new())));
    }

    #[test]
    fn test_get_struct_tokens_with_fields() {
        let raw_node = create_test_node_with_fields(
            "test_node",
            vec![(
                "test_field",
                FieldDefinition {
                    types: vec![TypeDefinition {
                        type_name: "test_type".to_string(),
                        named: true,
                    }],
                    multiple: false,
                    required: true,
                },
            )],
        );
        let language = get_language_no_nodes();
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(&node.get_struct_tokens(&HashSet::new())));
    }

    #[test]
    fn test_get_struct_tokens_complex() {
        let raw_node = create_test_node_with_fields(
            "test_node",
            vec![
                (
                    "required_field",
                    FieldDefinition {
                        types: vec![TypeDefinition {
                            type_name: "test_type".to_string(),
                            named: true,
                        }],
                        multiple: false,
                        required: true,
                    },
                ),
                (
                    "optional_field",
                    FieldDefinition {
                        types: vec![TypeDefinition {
                            type_name: "test_type".to_string(),
                            named: true,
                        }],
                        multiple: false,
                        required: false,
                    },
                ),
                (
                    "multiple_field",
                    FieldDefinition {
                        types: vec![TypeDefinition {
                            type_name: "test_type".to_string(),
                            named: true,
                        }],
                        multiple: true,
                        required: true,
                    },
                ),
            ],
        );
        let nodes = vec![raw_node.clone()];
        let language = get_language(nodes);
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(&node.get_struct_tokens(&HashSet::new())));
    }

    #[test_log::test]
    fn test_get_struct_tokens_with_children() {
        let raw_node =
            create_test_node_with_children("test_node", vec!["child_type_a", "child_type_b"]);
        let language = get_language_no_nodes();
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(&node.get_struct_tokens(&HashSet::new())));
    }

    #[test]
    fn test_get_struct_tokens_with_single_child_type() {
        let raw_node = create_test_node_with_children("test_node", vec!["child_type"]);
        let language = get_language_no_nodes();
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(&node.get_struct_tokens(&HashSet::new())));
    }

    #[test]
    fn test_get_trait_implementations() {
        let raw_node = create_test_node("test_node");
        let language = get_language_no_nodes();
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(
            &node.get_trait_implementations(&HashSet::new())
        ));
    }

    #[test]
    fn test_get_children_field_impl() {
        let raw_node = create_test_node_with_fields(
            "test_node",
            vec![(
                "test_field",
                FieldDefinition {
                    types: vec![TypeDefinition {
                        type_name: "test_type".to_string(),
                        named: true,
                    }],
                    multiple: false,
                    required: true,
                },
            )],
        );
        let language = get_language_no_nodes();
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(&node.get_children_field_impl()));
    }

    #[test]
    fn test_get_children_by_field_name_impl() {
        let raw_node = create_test_node_with_fields(
            "test_node",
            vec![(
                "test_field",
                FieldDefinition {
                    types: vec![TypeDefinition {
                        type_name: "test_type".to_string(),
                        named: true,
                    }],
                    multiple: false,
                    required: true,
                },
            )],
        );
        let language = get_language_no_nodes();
        let node = Node::new(Arc::new(raw_node), &language, Config::default());
        insta::assert_debug_snapshot!(snapshot_tokens(&node.get_children_by_field_name_impl()));
    }
}
