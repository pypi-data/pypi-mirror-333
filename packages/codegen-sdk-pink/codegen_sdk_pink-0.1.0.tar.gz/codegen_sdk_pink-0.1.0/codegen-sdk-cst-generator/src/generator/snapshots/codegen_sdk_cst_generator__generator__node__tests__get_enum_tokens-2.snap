---
source: codegen-sdk-cst-generator/src/generator/node.rs
expression: snapshot_tokens(&tokens)
---
# [subenum (subenum)] Test (Test < 'db1 >)
